lib/pbench/agent/tool_meister.py

```diff
@@ -897,27 +887,26 @@ def __enter__(self):
                 )
                 # Note - each ".install()" method runs sequentially.
                 res = tool.install()
-            except Exception:
+            except Exception as exc:
                 self.logger.exception("Failed to run tool %s install check", name)
-                res = InstallationResult(returncode=-42, output="internal-error")
+                res = InstallationResult(
+                    returncode=-42, output=f"internal-error: {exc}"
+                )
             # Record the result of the tool installation check so it can be
             # reported back to the Tool Data Sink.
             tool_installs[name] = res
-            if res.returncode != 0:
-                self.logger.debug("Recording failed tool, %s", name)
-                self._failed_tools[name] = self._tools[name]
-                del self._tools[name]
-            else:
+            if res.returncode == 0:
```

Comment:
Throwing away the `_failed_tools` because it's not actually used yet? Hmm.

Reply:
Figured it was better to _not_ perform a `del` of `self._params.tools` inside the loop where it was being used as part of the loop control and just keep track of the tools actually in use since `self._failed_tools` was never referenced after it was constructed.

Reply:
@portante, that's generally wise -- iterating over things like `dict`'s is not guaranteed to work if items are added or deleted during the iteration.  However, this loop is not actually iterating over `tools` -- it's iterating over a _list_ (i.e, a **copy**) of the items in `tools`, as provided by `sorted()` -- so we actually are free to modify `tools` here...if we want to.

Reply:
Yes, that is true, but that distinction, while correct, is still tenuous as the principle is useful regardless.

And I think the change is an improvement worth making.